### PR TITLE
mkhtml.py: get date from gitlog

### PR DESCRIPTION
**Pull request by ninsbl**
Trying to get the latest change and commit from gitlog.
Requires git and is not yet tested more broadly.
Please test. Feedback on HTML formating is very welcome!

### CODE REVIEW

File: utils/mkhtml.py [hunk 1]
```diff
@@ -68,6 +69,31 @@ def decode(bytes_):
...
+    cwd = os.getcwd()
+    # If git is not available we might try API but remember requests limit there
+    # trunk_api_url = f"https://api.github.com/repos/osgeo/grass/commits?path={{path}}&page=1&per_page=1
+    # addons_uapi_url = f"https://api.github.com/repos/osgeo/grass/commits?path={{path}}&page=1&per_page=1
+    # commits = urllib.request.urlopen(trunk_api_url)
+    # json.loads(commits.read().decode())
+    try:
+        os.chdir(src_dir)
+        gitlog = decode(check_output(["git", "log", "-1"])).split("\n")
+        print("src_dir: ", src_dir, "gitlog: ", gitlog)
```
Comment (nilason) [inline utils/mkhtml.py:83]: No need for this print here, I suppose debugging remnant.

**Comment by neteler** [comment]
I have tested locally and get the same error as shown in the "GCC C/C++..." pipeline:

```
Traceback (most recent call last):
  File "/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/utils/g.html2man.py", line 39, in main
    p.feed(line)
  File "/usr/lib/python3.8/html/parser.py", line 111, in feed
    self.goahead(0)
  File "/usr/lib/python3.8/html/parser.py", line 171, in goahead
    k = self.parse_starttag(i)
  File "/usr/lib/python3.8/html/parser.py", line 345, in parse_starttag
    self.handle_starttag(tag, attrs)
  File "/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/utils/ghtml.py", line 281, in handle_starttag
    while not self.is_allowed(tag):
  File "/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/utils/ghtml.py", line 277, in is_allowed
    return tag not in self.excluded and tag in allowed[self.top()]
KeyError: None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/utils/g.html2man.py", line 74, in <module>
    main()
  File "/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/utils/g.html2man.py", line 43, in main
    % (infile, err.lineno, err.offset, err.msg)
AttributeError: 'KeyError' object has no attribute 'lineno'
```

**Comment by tmszi** [comment]
> Trying to get the latest change and commit from gitlog. Requires git and is not yet tested more broadly. Please test. Feedback on HTML formating is very welcome!

Great idea, I looked at your code and it required a deeper analysis of the problem, I made changes in this commit https://github.com/OSGeo/grass/pull/2100/commits/83c3d7df329efe43892c32bbb0c31024084f370c and  I tested it, when compiling GRASS GIS the git local repo is used to get the commit and when installing the addon (via g.extension module) it is used to get the commit remote GitHub API server.

**Comment by ninsbl** [comment]
Seems we worked on this in parallel. Here is my approach for Addons and modules that are not part of the official addon repo:
https://github.com/ninsbl/grass/tree/mkhtml_git_addon
Your approach seems much more thorough, though, so i just pushed it to another branch, in case there is something wort transfering...

**Comment by ninsbl** [comment]
Seems addons that are not part of the official repo (e.g. local unversjoned addons, addons on gitlab, bitbucket, ...) need to be handled specificly:
https://github.com/OSGeo/grass/runs/4834443370?check_suite_focus=true#step:10:157
I would suggest to have a (one) fallback solution if no commit can be identified. So, for now it should be OK to not support all possible codesharing APIs (gitlab, bitbucket, gitea, ...))...
gitlab as another major platform may be added later (?)...

**Comment by tmszi** [comment]
> Seems addons that are not part of the official repo (e.g. local unversjoned addons, addons on gitlab, bitbucket, ...) need to be handled specificly: https://github.com/OSGeo/grass/runs/4834443370?check_suite_focus=true#step:10:157 I would suggest to have a (one) fallback solution if no commit can be identified.

Yes you are right I fixed it. 

>  So, for now it should be OK to not support all possible codesharing APIs (gitlab, bitbucket, gitea, ...))... gitlab as another major platform may be added later (?)...

Currently only official addons are supported, if addon is not official addon, commit and commit date in addon manual page will has unknown value.

File: utils/mkhtml.py [hunk 2]
```diff
@@ -121,6 +279,9 @@ def decode(bytes_):
   <a href="${URL_SOURCE}">${PGM} source code</a>
   (<a href="${URL_LOG}">history</a>)
 </p>
+<p>
+  Latest change: ${COMMIT_DATE} in commit: ${COMMIT}
```
Comment (ninsbl) [inline utils/mkhtml.py:283]: In https://github.com/ninsbl/grass/blob/e47c3673df97eaddcfa460a7a40a601b5101fc48/utils/mkhtml.py#L530-L536 I handled date from commit and file modification separately, maybe worth considering (e.g. for addons). Could be useful info to have in the manual, when it was installed... Not very important though...

Comment (tmszi) [reply utils/mkhtml.py]: Yes I agree. I have already implemented it https://github.com/OSGeo/grass/pull/2100/commits/d3f7e81faf7dc7d6db22048a5310f3b4c3635480.

**Comment by ninsbl** [comment]
Now all tests pass. Nice work, @tmszi , as usual. From my point this can be merged...

**Comment by tmszi** [comment]
> Now all tests pass. Nice work, @tmszi , as usual. From my point this can be merged...

I will have to test it as a simulation of compiling all add-ons running on our osgeo lxd container server.

**Comment by neteler** [comment]
I am +1 for merging, as patching our production system is always rather tricky...

**Comment by tmszi** [comment]
If user install addons via e.g. `g.extension i.sentinel proxy="http=<value>"` module and use proxy to connect to internet [mkhtml.py](https://github.com/OSGeo/grass/pull/2100/commits/83c3d7df329efe43892c32bbb0c31024084f370c#diff-3e1684c5c5d40b273b6488a9b5a5558f556d2bcf2973ba5106b6125e01aa6959R108) script have to use proxy settings too (exchange value between `g.extension` and mkhmtl.py script via global shell variable `GRASS_PROXY`).  I added global shell variable `GRASS_PROXY` with explanation among other shell variables in manual page https://github.com/OSGeo/grass/pull/2100/commits/d1437bb682b8d3a69c5391ce77d2a2f53229c03a.

**Comment by tmszi** [comment]
I successfully tested the simulation of the compilation of all add-ons on the osgeo lxd container server.

**Comment by neteler** [comment]
> I am +1 for merging, as patching our production system is always rather tricky...

FYI: I have enforced the compilation of binaries / addons on grass.osgeo.org. Hence, the improvements through this PR should now be visible.

**Review by tmszi — APPROVED** [review]

**Comment by tmszi** [comment]
> > I am +1 for merging, as patching our production system is always rather tricky...
> 
> FYI: I have enforced the compilation of binaries / addons on grass.osgeo.org. Hence, the improvements through this PR ~should now be visible.~ - nonsense as it hasn't been merged :D Sorry for the confusion.

It's already merged.

**Comment by wenzeslaus** [comment]
Similarly to another change in the past, the straight backport of this breaks Travis CI because that one builds with old Python. See the old discussion here: #2070.

```
VERSION_NUMBER=7.8.7dev VERSION_DATE=2022 MODULE_TOPDIR=../.. \
        python3 /home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/tools/mkhtml.py cairodriver > /home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/docs/html/cairodriver.html
  File "/home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/tools/mkhtml.py", line 74
    addons_url = f"https://github.com/OSGeo/grass-addons/tree/grass{major}/"
                                                                           ^
SyntaxError: invalid syntax
```

**Comment by tmszi** [comment]
> Similarly to another change in the past, the straight backport of this breaks Travis CI because that one builds with old Python. See the old discussion here: #2070.
> 
> ```
> VERSION_NUMBER=7.8.7dev VERSION_DATE=2022 MODULE_TOPDIR=../.. \
>         python3 /home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/tools/mkhtml.py cairodriver > /home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/docs/html/cairodriver.html
>   File "/home/travis/build/OSGeo/grass/dist.x86_64-pc-linux-gnu/tools/mkhtml.py", line 74
>     addons_url = f"https://github.com/OSGeo/grass-addons/tree/grass{major}/"
>                                                                            ^
> SyntaxError: invalid syntax
> ```

My mistake that originate during git cherry-picking.

**Comment by tmszi** [comment]
@neteler Before testing on the osgeo server, I still need to merge these two PRs #2110 #2111.

**Pull request merged**